thalassa/api.py

```diff
@@ -65,24 +65,102 @@ def get_elevation_dmap(trimesh: gv.TriMesh, show_grid: bool = False) -> hv.Overl
-    hd=hv.DynamicMap(get_timeseries_xy,streams=[PointerXY(x=0,y=0,source=source)])
-    hd=hd.opts(color='k',line_width=2,line_dash='dotted')
-
-    #points
-    hp=hv.HoloMap({'P_{}'.format(i+1):hv.Points(xy).opts(color='r') for i,xy in enumerate(xys)})
-
-    return hd*hs.overlay(),hp.overlay()
+#----------------------------------------------------------------------------------------
+#time series
+#----------------------------------------------------------------------------------------
+#define a class to store data related to time series points
+class timeseries_data:
+      def __init__(self):
+         self.init=False
+      def clear(self):
+          self.init=False
+
+#function for extracting time series@(x,y) from data
+def extract_timeseries(x,y,sx,sy,data):
+    dist=abs(sx+1j*sy-x-1j*y)
+    mdist=dist.min()
+    nid=np.nonzero(dist==mdist)[0][0]
+    mdata=data['elev'].data[:,nid].copy()
+    return mdist,mdata
+
+#function to dynamically add or remove pts by double clicking on the map
+def add_remove_pts(x,y,data,dataset,fmt):
+    if fmt=='add pts':
+       if len(data.xys)==0:
+          mdist,mdata=extract_timeseries(x,y,data.sx,data.sy,dataset)
+          hcurve=hv.Curve((data.time,mdata),'time','elevation')
+          if mdist<=data.mdist:
+             data.xys.append((x,y))
+             data.elev.append(mdata)
+             data.curve.append(hcurve)
+       else:
+          if data.xys[-1][0]!=x and data.xys[-1][1]!=y:
+             mdist,mdata=extract_timeseries(x,y,data.sx,data.sy,dataset)
+             hcurve=hv.Curve((data.time,mdata),'time','elevation')
+             if mdist<=data.mdist:
+                data.xys.append((x,y))
+                data.elev.append(mdata)
+                data.curve.append(hcurve)
+    elif fmt=='remove pts':
+       if len(data.xys)>0:
+          xys=np.array(data.xys)
+          dist=abs(xys[:,0]+1j*xys[:,1]-x-1j*y)
+          mdist=dist.min()
+          if mdist<=data.mdist:
+             nid=np.nonzero(dist==mdist)[0][0]
+             data.xys=[k for i,k in enumerate(data.xys) if i!=nid]
+             data.elev=[k for i,k in enumerate(data.elev) if i!=nid]
+             data.curve=[k for i,k in enumerate(data.curve) if i!=nid]
+    else:
+       pass
+
+#get time series plots
+def get_timeseries(source,data,dataset,ymin,ymax,fmt):
+    #initialize timeseries_data
+    if data.init is False:
+       #find the maximum side length
+       x,y=dataset['SCHISM_hgrid_node_x'].data,dataset['SCHISM_hgrid_node_y'].data
+       e1,e2,e3=dataset['SCHISM_hgrid_face_nodes'].data.T
+       s1=abs((x[e1]-x[e2])+1j*(y[e1]-y[e2])).max()
+       s2=abs((x[e2]-x[e3])+1j*(y[e2]-y[e3])).max()
+       s3=abs((x[e3]-x[e1])+1j*(y[e3]-y[e1])).max()
+
+       #save data
+       data.sx, data.sy, data.x0, data.y0  = x, y, x.mean(), y.mean()
+       data.mdist=np.max([s1,s2,s3])
+       data.time=dataset['time'].data
+       data.xys=[]
+       data.elev=[]
+       data.curve=[]
+       data.init=True
```

Comment:
I guess that here you are trying to "reset" the `timeseries_data` instance to a "vanilla" state. 

Nevertheless, I don't really like injecting attributes dynamically to "data containers". Principle of least surprise etc. I mean, when you read the code and you encounter the definition of `timeseries_data` class, how will you know that its instances are supposed to have all these extras attributes? how will you know what are their default values? How do you know  that can be reset, too?

These attributes and their default values should be defined in `timeseries_data.__init__()`. Resetting them to their default values could probably be done in `timeseries_data.clear()`.

Reply:
I will think about this. My original purpose is to make the code efficient and easy to maintain.